cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/WorkflowHttpHandler.java

```diff
@@ -345,6 +346,34 @@ private WorkflowToken getWorkflowToken(String namespaceId, String appName, Strin
   }
 
   @GET
+  @Path("/apps/{app-id}/workflows/{workflow-id}/runs/{run-id}/nodestates")
+  public void getWorkflowNodeStates(HttpRequest request, HttpResponder responder,
+                                       @PathParam("namespace-id") String namespaceId,
+                                       @PathParam("app-id") String applicationId,
+                                       @PathParam("workflow-id") String workflowId,
+                                       @PathParam("run-id") String runId)
+    throws NotFoundException, DatasetManagementException {
+    ApplicationId appId = new ApplicationId(namespaceId, applicationId);
+    ApplicationSpecification appSpec = store.getApplication(appId.toId());
+    if (appSpec == null) {
+      throw new ApplicationNotFoundException(appId.toId());
+    }
+
+    WorkflowSpecification workflowSpec = appSpec.getWorkflows().get(workflowId);
+    ProgramRunId workflowRunId = new ProgramRunId(namespaceId, applicationId, ProgramType.WORKFLOW, workflowId, runId);
+    if (workflowSpec == null) {
+      throw new ProgramNotFoundException(workflowRunId.getParent().toId());
+    }
+
+    if (store.getRun(workflowRunId.getParent().toId(), runId) == null) {
```

Comment:
store should really throw ``NotFoundException``. We do this everywhere. I guess there is a jira already.